Adopt UIAsyncTextInteraction behind the "Async UIKit Interactions" setting

whsieh opened the pull request:
#### d23335f7ed29a16f0efa731cfdfd4cde398b6a59
<pre>
Adopt UIAsyncTextInteraction behind the &quot;Async UIKit Interactions&quot; setting
<a href="https://bugs.webkit.org/show_bug.cgi?id=264020">https://bugs.webkit.org/show_bug.cgi?id=264020</a>

Reviewed by Megan Gardner.

Replace the existing `_textInteractionAssistant` in `WKContentView` with a `_textInteractionWrapper`
object instead, which wraps either a `UIWKTextInteractionAssistant` (in the case where the &quot;Async
UIKit Interactions&quot; feature flag is off), or a `UIAsyncTextInteraction` (in the case where the
feature flag is on).

For the legacy codepath, all methods on this new wrapper object correspond directly to the existing
methods already on `UIWKTextInteractionAssistant`, so there&apos;s no change in behavior; however, in the
case where we&apos;re using the new `UIAsyncTextInteraction`, only a subset of the methods have been
implemented so far, so we only plumb calls into `WKTextInteractionWrapper` through in those cases.

As we continue iterating on the modern async text interaction, we&apos;ll implement the remainder of
these methods which are currently no-ops when using `UIAsyncTextInteraction`.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _didExitStableState]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView shouldUseAsyncInteractions]):
(-[WKContentView _shouldUseUIContextMenuAsyncConfiguration]):
(-[WKContentView _shouldUseTextCursorDragAnimator]):
(-[WKContentView _updateRuntimeProtocolConformanceIfNeeded]):

Additionally declare conformance to `UIAsyncTextInputClient`, which extends from `UIAsyncTextInput`
and provides some additional functionality around text input via dictation (and more areas in the
near future).

(-[WKContentView cleanUpInteraction]):
(-[WKContentView becomeFirstResponderForWebView]):
(-[WKContentView endEditingAndUpdateFocusAppearanceWithReason:]):
(-[WKContentView _scrollingNodeScrollingWillBegin:]):
(-[WKContentView _scrollingNodeScrollingDidEnd:]):
(-[WKContentView setUpTextSelectionAssistant]):

Initialize a `WKTextInteractionWrapper` instead of a `UIWKTextInteractionAssistant`; this object
internally wraps either a legacy `UIWKTextInteractionAssistant`, or a new `UIAsyncTextInteraction`.

(-[WKContentView _willStartScrollingOrZooming]):
(-[WKContentView _didEndScrollingOrZooming]):
(-[WKContentView _lookupForWebView:]):
(-[WKContentView _shareForWebView:]):
(-[WKContentView _translateForWebView:]):
(-[WKContentView _addShortcutForWebView:]):
(-[WKContentView _promptForReplaceForWebView:]):
(-[WKContentView _transliterateChineseForWebView:]):
(-[WKContentView tintColorDidChange]):
(-[WKContentView selectForWebView:]):
(-[WKContentView selectAllForWebView:]):
(-[WKContentView _showDictionary:]):
(selectionChangedWithGesture):
(selectionChangedWithTouch):
(-[WKContentView changeSelectionWithGestureAt:withGesture:withState:withFlags:]):
(-[WKContentView changeSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):
(-[WKContentView changeSelectionWithTouchesFrom:to:withGesture:withState:]):
(-[WKContentView interactionAssistant]):
(-[WKContentView _showKeyboard]):
(-[WKContentView _hideKeyboard]):
(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):
(-[WKContentView _updateChangedSelection:]):
(-[WKContentView prepareSelectionForContextMenuWithLocationInView:completionHandler:]):
(-[WKContentView _startSuppressingSelectionAssistantForReason:]):
(-[WKContentView _stopSuppressingSelectionAssistantForReason:]):
(-[WKContentView _dragInteractionClass]):
(-[WKContentView _restoreEditMenuIfNeeded]):
(-[WKContentView dragInteraction:willAnimateLiftWithAnimator:session:]):
(-[WKContentView textInteractionAssistant]):
(-[WKContentView contextMenuInteraction]):
(-[WKContentView _registerPreview]):

Mechanically replace `_textInteractionAssistant` with `_textInteractionWrapper` in the content view.

* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.h: Added.
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.mm: Added.
(SOFT_LINK_CLASS_OPTIONAL):
(-[WKTextInteractionWrapper initWithView:]):
(-[WKTextInteractionWrapper textInteractionAssistant]):
(-[WKTextInteractionWrapper activateSelection]):
(-[WKTextInteractionWrapper deactivateSelection]):
(-[WKTextInteractionWrapper didEndScrollingOverflow]):
(-[WKTextInteractionWrapper selectionChanged]):
(-[WKTextInteractionWrapper setGestureRecognizers]):
(-[WKTextInteractionWrapper willStartScrollingOverflow]):
(-[WKTextInteractionWrapper selectionChangedWithGestureAt:withGesture:withState:withFlags:]):
(-[WKTextInteractionWrapper showDictionaryFor:fromRect:]):
(-[WKTextInteractionWrapper selectionChangedWithTouchAt:withSelectionTouch:withFlags:]):
(-[WKTextInteractionWrapper lookup:withRange:fromRect:]):
(-[WKTextInteractionWrapper showShareSheetFor:fromRect:]):
(-[WKTextInteractionWrapper showTextServiceFor:fromRect:]):
(-[WKTextInteractionWrapper scheduleReplacementsForText:]):
(-[WKTextInteractionWrapper scheduleChineseTransliterationForText:]):
(-[WKTextInteractionWrapper willStartScrollingOrZooming]):
(-[WKTextInteractionWrapper didEndScrollingOrZooming]):
(-[WKTextInteractionWrapper selectWord]):
(-[WKTextInteractionWrapper selectAll:]):
(-[WKTextInteractionWrapper contextMenuInteraction]):
(-[WKTextInteractionWrapper setExternalContextMenuInteractionDelegate:]):
(-[WKTextInteractionWrapper translate:fromRect:]):

Plumb from the text interaction wrapper to the inner interaction assistant; we also call into the
`_textInteractionWrapper` in only the methods that currently have parity with
`UIWKTextInteractionAssistant`.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/270136@main">https://commits.webkit.org/270136@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e328b33e24ddf5fe68da13cf670f429a5b43089

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24625 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3174 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25882 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26748 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22624 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/4836 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/610 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23002 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24870 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2235 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21267 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27331 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1956 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28378 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22466 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22535 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26180 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1879 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/204 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3185 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5905 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2334 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2247 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

megangardner reviewed (approved):
Great wrapper to help this transition!

whsieh committed:
Adopt UIAsyncTextInteraction behind the "Async UIKit Interactions" setting
https://bugs.webkit.org/show_bug.cgi?id=264020

Reviewed by Megan Gardner.

Replace the existing `_textInteractionAssistant` in `WKContentView` with a `_textInteractionWrapper`
object instead, which wraps either a `UIWKTextInteractionAssistant` (in the case where the "Async
UIKit Interactions" feature flag is off), or a `UIAsyncTextInteraction` (in the case where the
feature flag is on).

For the legacy codepath, all methods on this new wrapper object correspond directly to the existing
methods already on `UIWKTextInteractionAssistant`, so there's no change in behavior; however, in the
case where we're using the new `UIAsyncTextInteraction`, only a subset of the methods have been
implemented so far, so we only plumb calls into `WKTextInteractionWrapper` through in those cases.

As we continue iterating on the modern async text interaction, we'll implement the remainder of
these methods which are currently no-ops when using `UIAsyncTextInteraction`.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _didExitStableState]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView shouldUseAsyncInteractions]):
(-[WKContentView _shouldUseUIContextMenuAsyncConfiguration]):
(-[WKContentView _shouldUseTextCursorDragAnimator]):
(-[WKContentView _updateRuntimeProtocolConformanceIfNeeded]):

Additionally declare conformance to `UIAsyncTextInputClient`, which extends from `UIAsyncTextInput`
and provides some additional functionality around text input via dictation (and more areas in the
near future).

(-[WKContentView cleanUpInteraction]):
(-[WKContentView becomeFirstResponderForWebView]):
(-[WKContentView endEditingAndUpdateFocusAppearanceWithReason:]):
(-[WKContentView _scrollingNodeScrollingWillBegin:]):
(-[WKContentView _scrollingNodeScrollingDidEnd:]):
(-[WKContentView setUpTextSelectionAssistant]):

Initialize a `WKTextInteractionWrapper` instead of a `UIWKTextInteractionAssistant`; this object
internally wraps either a legacy `UIWKTextInteractionAssistant`, or a new `UIAsyncTextInteraction`.

(-[WKContentView _willStartScrollingOrZooming]):
(-[WKContentView _didEndScrollingOrZooming]):
(-[WKContentView _lookupForWebView:]):
(-[WKContentView _shareForWebView:]):
(-[WKContentView _translateForWebView:]):
(-[WKContentView _addShortcutForWebView:]):
(-[WKContentView _promptForReplaceForWebView:]):
(-[WKContentView _transliterateChineseForWebView:]):
(-[WKContentView tintColorDidChange]):
(-[WKContentView selectForWebView:]):
(-[WKContentView selectAllForWebView:]):
(-[WKContentView _showDictionary:]):
(selectionChangedWithGesture):
(selectionChangedWithTouch):
(-[WKContentView changeSelectionWithGestureAt:withGesture:withState:withFlags:]):
(-[WKContentView changeSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):
(-[WKContentView changeSelectionWithTouchesFrom:to:withGesture:withState:]):
(-[WKContentView interactionAssistant]):
(-[WKContentView _showKeyboard]):
(-[WKContentView _hideKeyboard]):
(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):
(-[WKContentView _updateChangedSelection:]):
(-[WKContentView prepareSelectionForContextMenuWithLocationInView:completionHandler:]):
(-[WKContentView _startSuppressingSelectionAssistantForReason:]):
(-[WKContentView _stopSuppressingSelectionAssistantForReason:]):
(-[WKContentView _dragInteractionClass]):
(-[WKContentView _restoreEditMenuIfNeeded]):
(-[WKContentView dragInteraction:willAnimateLiftWithAnimator:session:]):
(-[WKContentView textInteractionAssistant]):
(-[WKContentView contextMenuInteraction]):
(-[WKContentView _registerPreview]):

Mechanically replace `_textInteractionAssistant` with `_textInteractionWrapper` in the content view.

* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.h: Added.
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.mm: Added.
(SOFT_LINK_CLASS_OPTIONAL):
(-[WKTextInteractionWrapper initWithView:]):
(-[WKTextInteractionWrapper textInteractionAssistant]):
(-[WKTextInteractionWrapper activateSelection]):
(-[WKTextInteractionWrapper deactivateSelection]):
(-[WKTextInteractionWrapper didEndScrollingOverflow]):
(-[WKTextInteractionWrapper selectionChanged]):
(-[WKTextInteractionWrapper setGestureRecognizers]):
(-[WKTextInteractionWrapper willStartScrollingOverflow]):
(-[WKTextInteractionWrapper selectionChangedWithGestureAt:withGesture:withState:withFlags:]):
(-[WKTextInteractionWrapper showDictionaryFor:fromRect:]):
(-[WKTextInteractionWrapper selectionChangedWithTouchAt:withSelectionTouch:withFlags:]):
(-[WKTextInteractionWrapper lookup:withRange:fromRect:]):
(-[WKTextInteractionWrapper showShareSheetFor:fromRect:]):
(-[WKTextInteractionWrapper showTextServiceFor:fromRect:]):
(-[WKTextInteractionWrapper scheduleReplacementsForText:]):
(-[WKTextInteractionWrapper scheduleChineseTransliterationForText:]):
(-[WKTextInteractionWrapper willStartScrollingOrZooming]):
(-[WKTextInteractionWrapper didEndScrollingOrZooming]):
(-[WKTextInteractionWrapper selectWord]):
(-[WKTextInteractionWrapper selectAll:]):
(-[WKTextInteractionWrapper contextMenuInteraction]):
(-[WKTextInteractionWrapper setExternalContextMenuInteractionDelegate:]):
(-[WKTextInteractionWrapper translate:fromRect:]):

Plumb from the text interaction wrapper to the inner interaction assistant; we also call into the
`_textInteractionWrapper` in only the methods that currently have parity with
`UIWKTextInteractionAssistant`.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/270136@main
